src/xtopcom/xdata/src/xtop_action_generator.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2017-2021 Telos Foundation & contributors
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "xdata/xtop_action_generator.h"
+
+#include "xbasic/xmemory.hpp"
+#include "xcommon/xaddress.h"
+#include "xdata/xconsensus_action.h"
+#include "xdata/xerror/xerror.h"
+
+#include <cassert>
+
+NS_BEG2(top, contract_runtime)
+
+std::unique_ptr<data::xbasic_top_action_t const> xtop_action_generator::generate(xobject_ptr_t<data::xcons_transaction_t> const & tx) {
+    std::error_code ec;
+    auto r = generate(tx, ec);
+    top::error::throw_error(ec);
+    return r;
+}
+
+std::unique_ptr<data::xbasic_top_action_t const> xtop_action_generator::generate(xobject_ptr_t<data::xcons_transaction_t> const & tx, std::error_code & ec) {
+    common::xaccount_address_t const target_address{ tx->get_transaction()->get_target_addr() };
+    
+    // eth deploy code
+    if (target_address.empty()) {
```

Comment:
for deploying code, it is `empty` or is `0x00...00`?

Reply:
both ok, it`s decided by how rpc wrap the cons_tx.